### ISSUE TITLE
instanceof doesn't work

### ISSUE THREAD

**Issue by zackbloom** [HubSpot/mixen#3]
AFAIK it is not possible to make the native instanceof work.  Per [the spec](http://ecma262-5.com/ELS5_HTML.htm#Section_15.3.4.5.3), instanceof traverses the prototype chain of the object, looking for an object with the prototype of the function in question.  Unfortunately, a function can only have one prototype, so the minute you wanted to use a mixin in more than one mixen, you'd be hosed.

It would be possible however to create a `Mixen.instanceof` method which would work.

**Comment by dpehrson** [comment]
As the person who initially reported this, I just wanted to chime in.

This is used in some of our larger JS applications for doing type/interface checking of function arguments. Technically you could use duck typing to work around the issue, but I just personally enjoy being to query the object to ask "Are you a SomeMixiner?" before moving on.

While being able to use `instanceof` would be ideal, falling back to something a helper works. I think there are a few combinations of ways to do this that depends on how much we want to "pollute" the mixed-in object and what we expect.

To me there are two main questions:
1. Do we want to add a helper that **only** handles the question of whether an object has another object mixed or do we want a helper that handles mixins and class extensions? If the former the name of such a method should probably be something along the lines of `[mM]ixesIn`. If the latter it would probably make sense to name it something more along the lines of `instanceof`
2. Do we want to make it a static method of `Mixen` itself or pollute the general object space. I think that the latter is clearly better in terms of polite programming even if the former would be more syntactically sugary.

With that in mind, I think the best option is to provide a few static helper methods on Mixen that answer different queries:
1. `Boolean Mixen.mixinof(Class instance, Class klass)`: Returns true only if `instance` has incorporated `klass` as a mixin, but is not an instance of it directly or through extension.
2. `Boolean Mixen.instanceof(Class instance, Class klass)`: Returns true only if `instance` is an instance of `klass` directly or through extension, but not through incorporation via mixin. This is basically a straight wrapper for `instanceof` exposed simply for consistency. If people are going to have to query Mixen for these kinds of things, it might be good to do it for all queries.
3. `Boolean Mixen.???(Class instance, Class klass)`: Returns true if `instance` incorporates `klass` either because it is one, extends one, or mixed it in. I don't actually have a good name off the top of my head for this.

Thoughts?

Depending on my schedule, I might fork and poke around in the code to see if I can tinker if you don't get to it first.

**Comment by zackbloom** [comment]
I completely agree.

The names work for me, although I am a little concerned that `instanceof` being a reserved word might be a problem, we might want to switch to `isinstance`.  The name for #3 could just be `includes`.

One thought is that we want to provide a method which is a drop-in replacement for `instanceof` in that it can be used on classes and instances which have nothing to do with mixer (or mixers), and functions as close to how someone used to `instanceof` would (maybe #3?)

The other option is to bind static methods onto the created mixen class itself: `if MyMixen.includes(SomeMixin)`.  We're not going to want to pass that along to the instances though, so it's less applicable to this issue.

**Comment by dpehrson** [comment]
Yeah, I think 2 is really just a wrapper for `instanceof` solely so all class/mixin type comparison can be run through Mixen.

For naming, it would be possible to do some funky stuff with returned function references (rspec-style) to have more syntax-friendly public functions:

```
# Using a real-world example from an rspec-style JS test framework I am building...

# Mixin for things that can describe other things
class Describer
  describe: (name, block) => ...

# Mixen for things that can have contexts
class Contexter
  context: (name, block) => ...

# Mixen for things that can have examples
class Exampler
  it: (label, block) => ...

# Describe class can describe children, have contexts, or create examples.
class Describe extends Mixen(Contexter, Describer, Exampler)

# Context class can have sub-contexts, or create examples.
class Context extends Mixen(Contexter, Exampler)

# Fake class to show extension.
class RandomClass extends Context

## Now for fun querying...

Mixen.does(new Context()).mixin(Describer) # => false, it does not.
Mixen.does(new Context()).mixin(Exampler) # => true, it does.
Mixen.does(new Context()).mixin(Context) # => false, Context is a class, not a mixin.

Mixen.does(new RandomClass()).mixin(Describer) # => false, it does not because it's parent (Context) does not.
Mixen.does(new RandomClass()).mixin(Contexter) # => true, it does, because it's parent (Context) class does.

Mixen.does(new RandomClass()).extend(RandomClass) # => false, it does not extend it, it is it.
Mixen.does(new RandomClass()).extend(Context) # => true, it does.

Mixen.is(new RandomClass()).a(RandomClass) # => true, it is directly.
Mixen.is(new RandomClass()).a(Context) # => true, it is by extension.
Mixen.is(new RandomClass()).a(Describe) # => false, is not.
Mixen.is(new Context()).a(Contexter) # => true, by virtue of mixin.
```

Please note I am not officially proposing the above, I haven't even finalized my thoughts, but something along those lines would allow for fairly readable testing of various types of inclusion.

**Comment by zackbloom** [comment]
Not a full solution, but this is relevant: https://github.com/HubSpot/mixen/commit/5fe80d97dec3eea8eaed84db1665b4d7a57ac699
